File: baselib/src/main/java/com/example/baselib/di/module/AppModule.java
package com.example.baselib.di.module;

import android.app.Application;
import android.content.Context;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

@Module
public abstract class AppModule {

//    @Singleton
//    @Provides
//    static Gson provideGson(Application application, @Nullable GsonConfiguration configuration) {
//        GsonBuilder builder = new GsonBuilder();
//        if (configuration != null)
//            configuration.configGson(application, builder);
//        return builder.create();
//    }
//
//    /**
//     * 之前 {@link AppManager} 使用 Dagger 保证单例, 只能使用 {@link AppComponent#appManager()} 访问
//     * 现在直接将 AppManager 独立为单例类, 可以直接通过静态方法 {@link AppManager#getAppManager()} 访问, 更加方便
//     * 但为了不影响之前使用 {@link AppComponent#appManager()} 获取 {@link AppManager} 的项目, 所以暂时保留这种访问方式
//     *
//     * @param application
//     * @return
//     */
//    @Singleton
//    @Provides
//    static AppManager provideAppManager(Application application) {
//        return AppManager.getAppManager().init(application);
//    }
//
//    @Binds
//    abstract IRepositoryManager bindRepositoryManager(RepositoryManager repositoryManager);
//
//    @Singleton
//    @Provides
//    static Cache<String, Object> provideExtras(Cache.Factory cacheFactory) {
//        return cacheFactory.build(CacheType.EXTRAS);
//    }
//
//    @Binds
//    @Named("ActivityLifecycle")
//    abstract Application.ActivityLifecycleCallbacks bindActivityLifecycle(ActivityLifecycle activityLifecycle);
//
//    @Binds
//    @Named("ActivityLifecycleForRxLifecycle")
//    abstract Application.ActivityLifecycleCallbacks bindActivityLifecycleForRxLifecycle(ActivityLifecycleForRxLifecycle activityLifecycleForRxLifecycle);
//
//    @Binds
//    abstract FragmentManager.FragmentLifecycleCallbacks bindFragmentLifecycle(FragmentLifecycle fragmentLifecycle);
//
//    @Singleton
//    @Provides
//    static List<FragmentManager.FragmentLifecycleCallbacks> provideFragmentLifecycles() {
//        return new ArrayList<>();
//    }
//
//    public interface GsonConfiguration {
//        void configGson(@NonNull Context context, @NonNull GsonBuilder builder);
//    }
}

File: baselib/src/main/java/com/example/baselib/di/module/AppModule.kt
package com.example.baselib.di.module


import android.app.Application
import android.content.Context

import java.util.ArrayList

import javax.inject.Named
import javax.inject.Singleton
import androidx.fragment.app.FragmentManager
import com.example.baselib.integration.*
import com.example.baselib.integration.cache.Cache
import com.example.baselib.integration.cache.CacheType
import com.example.baselib.integration.lifecycle.ActivityLifecycleForRxLifecycle
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import dagger.Binds
import dagger.Module
import dagger.Provides

@Module
abstract class AppModule {

    @Binds
    internal abstract fun bindRepositoryManager(repositoryManager: RepositoryManager): IRepositoryManager

    @Binds
    @Named("ActivityLifecycle")
    internal abstract fun bindActivityLifecycle(activityLifecycle: ActivityLifecycle): Application.ActivityLifecycleCallbacks

    @Binds
    @Named("ActivityLifecycleForRxLifecycle")
    internal abstract fun bindActivityLifecycleForRxLifecycle(activityLifecycleForRxLifecycle: ActivityLifecycleForRxLifecycle): Application.ActivityLifecycleCallbacks

    @Binds
    internal abstract fun bindFragmentLifecycle(fragmentLifecycle: FragmentLifecycle): FragmentManager.FragmentLifecycleCallbacks

    interface GsonConfiguration {
        fun configGson(context: Context, builder: GsonBuilder)
    }

    companion object {

        @Singleton
        @Provides
        internal fun provideGson(application: Application, configuration: GsonConfiguration?): Gson {
            val builder = GsonBuilder()
            configuration?.configGson(application, builder)
            return builder.create()
        }

        /**
         * 之前 [AppManager] 使用 Dagger 保证单例, 只能使用 [@AppComponent#appManager()][@AppComponent.appManager] 访问
         * 现在直接将 AppManager 独立为单例类, 可以直接通过静态方法 [AppManager.getAppManager] 访问, 更加方便
         * 但为了不影响之前使用 [@AppComponent#appManager()][@AppComponent.appManager] 获取 [AppManager] 的项目, 所以暂时保留这种访问方式
         *
         * @param application
         * @return
         */
        @Singleton
        @Provides
        internal fun provideAppManager(application: Application): AppManager {
            return AppManager.getAppManager().init(application)
        }

        @Singleton
        @Provides
        internal fun provideExtras(cacheFactory: Cache.Factory): Cache<String, Any> {
            return cacheFactory.build(CacheType.EXTRAS) as Cache<String, Any>
        }

        @Singleton
        @Provides
        internal fun provideFragmentLifecycles(): List<FragmentManager.FragmentLifecycleCallbacks> {
            return ArrayList()
        }
    }
}
